### PR TITLE
Update input configuration.

### PR DESCRIPTION
**Pull request by josiah-tesfu**
Updated the input configuration. Configuration is initially set through the editor. Once the node is deployed, configuration can be set dynamically from input. If any parameters from the input message are empty, the previously set configurations are used.

### CODE REVIEW

File: edgepi-dac.js
Comment (sjpark608) [inline edgepi-dac.js:1]: I'm little confused. what's difference between initializenode() and initializenode().then(dac())?
It only has case checking when gain is true, what happens if the user wants to initialize it with gain false()?

File: README.md [hunk 1]
```diff
@@ -15,21 +15,35 @@ npm install @edgepi-cloud/node-red-edgepi-digital-dac
...
+'false' for gain disabled (off), 'true' for gain enabled (on).
+  - **channel** *number*: <br>
+
+Example input:
+```
+msg {
+  "payload": 5,
+  "gain": false,
+  "channel": 7
+}
```
Comment (farzadpanahi) [inline README.md]: payload should be something like "Voltage to write"

in the example have a float number like 2.5

**Review by farzadpanahi — CHANGES_REQUESTED** [review]

File: edgepi-dac.js [hunk 2]
```diff
@@ -2,61 +2,67 @@ module.exports = function (RED) {
...
-    } else {
-      node.status({
-        fill: "red",
-        shape: "ring",
-        text: "dac failed to initialize.",
+          if ((gain && voltage > 10) || voltage < 0 || (!gain && voltage > 5)) {
+            throw new Error(
+              "Voltage being written is outside the valid range."
+            );
+          }
```
Comment (naomiven) [inline edgepi-dac.js]: how does `gain && voltage` evaluate to a number?

File: edgepi-dac.html [hunk 3]
```diff
@@ -18,48 +18,46 @@
...
+        gain2.checked = true;
       }
 
       function updateEditor() {
         tcpField.style.display =
           configTransportSelect.value === "Network" ? "flex" : "none";
-        const selectedGain = gainOption1.checked ? "gain1" : "gain2";
-        voltageGain1.style.display = selectedGain === "gain1" ? "flex" : "none";
-        voltageGain2.style.display = selectedGain === "gain2" ? "flex" : "none";
+        const selectedGain = gain1.checked ? false : true;
```
Comment (jimmy121192) [inline edgepi-dac.html]: use this : const selectedGain = !gain1.checked; for this line and all the similar lines below

File: edgepi-dac.html [hunk 4]
```diff
@@ -18,48 +18,46 @@
...
+        voltageGain1.style.display = selectedGain === false ? "flex" : "none";
+        voltageGain2.style.display = selectedGain === true ? "flex" : "none";
       }
 
-      // Update and add event listener to update on user input
       updateEditor();
       document.getElementById(
-        this.gainOption === "gain1"
+        this.gain === false
```
Comment (jimmy121192) [inline edgepi-dac.html]: !this.gain ? "node-input-voltageGain1"
          : "node-input-voltageGain2"

File: edgepi-dac.js [hunk 5]
```diff
@@ -2,61 +2,67 @@ module.exports = function (RED) {
...
+    let voltage = config.voltage;
+    let gain = config.gain;
+    let channel = config.channel;
 
-    const dac = new rpc.DacService(transport);
+    initializeNode(config).then((dac) => {
+      node.on("input", async function (msg, send, done) {
+        node.status({ fill: "green", shape: "dot", text: "input received" });
+        try {
+          gain = msg.gain ?? gain;
```
Comment (jimmy121192) [inline edgepi-dac.js]: I have mentioned an issue with  the operator ?? in
https://github.com/EdgePi-Cloud/node-red-edgepi-thermocouple/pull/29#pullrequestreview-1798405803
Please take a look.

**Review by jimmy121192 — CHANGES_REQUESTED** [review]

File: edgepi-dac.js (hunk 2 above)
Comment (josiah-tesfu) [reply edgepi-dac.js]: Those are evaluated separately. Order of operations in that part: `gain`, `voltage > 10`, `gain && voltage > 10`.

File: edgepi-dac.js
Comment (josiah-tesfu) [reply edgepi-dac.js:1]: .then... executes synchronously. Calls initializeNode(), waits for completion, and then continues.

setDacGain() only needs to be called when gain is true. By default gain isn't set, so if gain is false nothing needs to be done.

**Review by naomiven — DISMISSED** [review]

**Review by jimmy121192 — DISMISSED** [review]

File: edgepi-dac.html [hunk 6]
```diff
@@ -18,48 +18,44 @@
...
     },
     oneditprepare: function () {
-      // Query html elements
       const configTransportSelect = document.getElementById(
         "node-input-transport"
       );
-      const gainOption1 = document.getElementById("node-input-gain1");
-      const gainOption2 = document.getElementById("node-input-gain2");
+      const gain1 = document.getElementById("node-input-gain1");
+      const gain2 = document.getElementById("node-input-gain2");
```
Comment (farzadpanahi) [inline edgepi-dac.html]: what is gain1 and gain2? I am confused : /

Comment (farzadpanahi) [inline edgepi-dac.html]: what is gain1 and gain2? I am confused : /

File: edgepi-dac.js [hunk 7]
```diff
@@ -2,61 +2,67 @@ module.exports = function (RED) {
...
-      node.status({ fill: "green", shape: "dot", text: "input received" });
       try {
-        if (config.gain2) {
-          console.log("before set gain");
+        const dac = new rpc.DacService(transport);
+        console.info("DAC node initialized on:", transport);
+        node.status({ fill: "green", shape: "ring", text: "dac initialized" });
+
+        if ((gain && voltage > 10) || voltage < 0 || (!gain && voltage > 5)) {
+          throw new Error("Voltage being written is outside the valid range.");
```
Comment (farzadpanahi) [inline edgepi-dac.js]: let's specify the valid ragne

**Review by farzadpanahi — CHANGES_REQUESTED** [review]

File: edgepi-dac.html (hunk 6 above)
Comment (jimmy121192) [reply edgepi-dac.html]: Updated to gainOn and gainOff

Comment (jimmy121192) [reply edgepi-dac.html]: Updated to gainOn and gainOff

File: edgepi-dac.js (hunk 7 above)
Comment (jimmy121192) [reply edgepi-dac.js]: updated

**Review by farzadpanahi — APPROVED** [review]

**Pull request merged**